airdrop-machine/sources/airdrop_machine.move

```diff
@@ -0,0 +1,290 @@
+        royalty_denominator: &mut Option<u64>, 
+        admin_addr: address
+    ): Option<Royalty> {
+        assert!(option::is_some(royalty_numerator) == option::is_some(royalty_denominator), error::invalid_argument(EROYALTY_CONFIGURATION_INVALID));
+        if (option::is_some(royalty_numerator) && option::is_some(royalty_denominator) && option::extract(royalty_numerator) != 0 && option::extract(royalty_denominator) != 0) {
+            option::some(royalty::create(option::extract(royalty_numerator), option::extract(royalty_denominator), admin_addr))
+        } else {
+            option::none()
+        }
+    }
+
+    fun configure_collection_and_token_properties(
+        admin: &signer,
+        collection: Object<Collection>,
+        mutable_collection_metadata: bool,
+        mutable_token_metadata: bool,
+        tokens_burnable_by_collection_owner: bool,
+        tokens_transferrable_by_collection_owner: bool,
+    ) {
+        collection_properties::set_mutable_description(admin, collection, mutable_collection_metadata);
+        collection_properties::set_mutable_uri(admin, collection, mutable_collection_metadata);
+        collection_properties::set_mutable_royalty(admin, collection, mutable_collection_metadata);
+        collection_properties::set_mutable_token_name(admin, collection, mutable_token_metadata);
+        collection_properties::set_mutable_token_properties(admin, collection, mutable_token_metadata);
+        collection_properties::set_mutable_token_description(admin, collection, mutable_token_metadata);
+        collection_properties::set_mutable_token_uri(admin, collection, mutable_token_metadata);
+        collection_properties::set_tokens_transferable_by_collection_owner(admin, collection, tokens_transferrable_by_collection_owner);
+        collection_properties::set_tokens_burnable_by_collection_owner(admin, collection, tokens_burnable_by_collection_owner);
+    }
+
+    fun assert_owner<T: key>(owner: address, object: Object<T>) {
+        assert!(object::owner(object) == owner, error::permission_denied(ENOT_OWNER));
+    }
+
+    inline fun collection_owner_signer(collection_config: &CollectionConfig): signer acquires CollectionConfig {
+        object::generate_signer_for_extending(&collection_config.extend_ref)
+    }
+
+    inline fun borrow(collection_config_object: Object<CollectionConfig>): &CollectionConfig acquires CollectionConfig {
+        borrow_global<CollectionConfig>(object::object_address(&collection_config_object))
+    }
+
+    inline fun borrow_mut(collection_config_object: Object<CollectionConfig>): &mut CollectionConfig acquires CollectionConfig {
+        borrow_global_mut<CollectionConfig>(object::object_address(&collection_config_object))
+    }
+
+    fun get_pseudo_random_index(length: u64): u64 {
+        let txn_hash = transaction_context::get_transaction_hash();
+        ((*vector::borrow(&txn_hash, 0) as u64) * 256u64) % length
+    }
```

Comment:
this is not working @igor-aptos . `generate_auid_address` returns the `address`, can we `vector::borrow` an address thou?

im using `get_transaction_hash` but it seems like vector only has a length of 1 cuz `*vector::borrow(&txn_hash, 1)` errors out. 


meanwhile, its still not working cuz unit tests all fail with:

```
│     ┌─ /Users/jillxu/.move/https___github_com_aptos-labs_aptos-core_git_mainnet/aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move:118:23
│     │
│ 118 │     public native fun string_concat<IntElement>(before: String, snapshot: &AggregatorSnapshot<IntElement>, after: String): AggregatorSnapshot<String>;
│     │                       ^^^^^^^^^^^^^
│     │                       │
│     │                       Test was not expected to error, but it aborted with code 196617 originating in the module 0000000000000000000000000000000000000000000000000000000000000001::aggregator_v2 rooted here
│     │                       In this function in 0x1::aggregator_v2
│ 
│ 
```